wdl/src/main/scala/wdl/command/ParameterCommandPart.scala

```diff
@@ -49,29 +45,45 @@ case class ParameterCommandPart(attributes: Map[String, String], expression: Wdl
 
     val lookup: String => WomValue = (s: String) => valueMapper(inputs.getOrElse(s, lookupDeclaration(s)))
 
-    val value = expression.evaluate(lookup, ParameterCommandPart.withPostProcesser(valueMapper, functions)) match {
+    // Note that evaluating this expression may have the side effect of causing a file to be created if `writeFile`
+    // is invoked. That file will be written to an "engine-relative" location which may be in cloud storage.
+    val evaluatedCommandPartExpression: ErrorOr[WomValue] = expression.evaluate(lookup, functions) match {
       case Success(v) => v match {
-        case WomOptionalValue(_, opt) => opt.getOrElse(defaultString)
-        case _ => v
+        case WomOptionalValue(_, opt) => opt.getOrElse(defaultString).validNel
+        case _ => v.validNel
       }
-      case Failure(OptionalNotSuppliedException(_)) => defaultString
-      case Failure(f) => throw new UnsupportedOperationException(s"Could not evaluate expression: ${expression.toWomString}", f)
+      case Failure(OptionalNotSuppliedException(_)) => defaultString.validNel
+      case Failure(_) => s"Could not evaluate expression: ${expression.toWomString}".invalidNel
     }
 
-    valueMapper(value) match {
-      case b: WomBoolean if attributes.contains("true") && attributes.contains("false") => if (b.value) attributes.get("true").head else attributes.get("false").head
-      case p: WomPrimitive => p.valueString
-      case a: WomArray if attributes.contains("sep") => a.value.map(_.valueString).mkString(attributes.get("sep").head)
-      case _: WomArray => throw new UnsupportedOperationException(s"Expression '${expression.toString}' evaluated to an Array but no 'sep' was specified")
-      case _ => throw new UnsupportedOperationException(s"Could not string-ify value: $value")
-    }
-  }
+    // Create the stringified version of the command and record any file created by `writeFile`.
```

Comment:
`"by 'writeFile"` => `"in the process"`?

Is there any chance CWL might like this to be in the WOM layer?

Reply:
I don't know but we can refactor if that turns out to be the case.